Base Code for Course Edit and Delete

spencerj411 opened the pull request:
As #29 _Course Deletion (8)_ is currently blocked, I created a new branch that contains base code for both #31 _Course Editing (5)_ (@Daniel-Jang) and #29 instead.
The idea was to render just one instance of the course table and have an 'Options' column where the user could choose to either edit or delete a course. 

Notes:
1. code contains placeholders (sample links and images)
2. the icons used were from http://webpublishing.anu.edu.au/web-style-guide/icons.php. However, we are expected to contact ANU styling first if we want to use their icons for the MVP.
3. ANU styling has no icons that could be used for 'Delete' (hence the weird choice of thumbs down). However, we could also contact them about this if we want.

![Capture](https://user-images.githubusercontent.com/24206502/56189655-d6061200-606b-11e9-9c02-50e032689a28.JPG)

spencerj411 committed:
Created base code for Edit and Delete (contains some temp links/images that act as placeholders).Changed [Add, Edit, Delete] tabs to [Add, Modify].

spencerj411 committed:
removed ?action=modify comment

pep8speaks commented:
Hello @spencerj411! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`cassdegrees/ui/views.py`](https://github.com/cass-degrees/CASS-Degrees-Code/blob/6757f073c713daa8897589327f49877287506e5e/cassdegrees/ui/views.py):

> [Line 105:32](https://github.com/cass-degrees/CASS-Degrees-Code/blob/6757f073c713daa8897589327f49877287506e5e/cassdegrees/ui/views.py#L105): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

JackSekoranja reviewed (changes requested):
I understand the reasoning behind using a 'thumbs down' picture, I don't think it's a good idea to have this in the final version. The layout of the pictures is also slightly different for me (See attached picture). I also think the newly created the newly created search function on the `/list/` page would be useful in this instance, and since they perform a very similar function, it may be better just to adapt that page and use it to find courses to edit/delete. Plus it means that we could have a consistent page for editing/removing existing courses, degrees, or subplans.

![Edit Format](https://user-images.githubusercontent.com/36946090/56206107-ebd8fe80-608e-11e9-85e1-24c0f7a74157.jpg)

Also, just a note. You could continue to work on the actual deletion of courses themselves, as that will take some time. The only area you are blocked on is searching for course dependencies in subplans

spencerj411 commented:
I should’ve been clearer about a few things:
1. The thumbs down icon was chosen temporarily just to see how 2 icons would look like in the Options column. We could either contact ANU styling about creating a new delete icon or we could just have them as text. Thoughts?
2. I have been working on the non-blocked criteria of Course Deletion but my thought process was to pr the overlapping work between delete and edit so Daniel could start working on edit concurrently.

Idk why the icons are positioned like that though, I used their css; I’ll have a look into it. I’ll also have a look at list.html and implement the search function once list.html changes have been merged.

JackSekoranja commented:
1. I think the preferred option here would be to find different pictures (outside ANU ones) for the edit and delete functions, or if we couldn't do that then just use text.
2. Fair enough

I also wasn't suggesting you implement the search function yourself, since it has a fair amount of overhead. I thought it might be easier to replace the edit form with the list form and have a single consistent one (Which would benefit code re-use). I was thinking we could separate the main menu in to 3-4 basic functions: 
- **Add New Template/Subplan/Course**
- **View Existing Template/Subplan/Course**
- **Edit Existing Template/Subplan/Course**
- **Output/Print Template**

Then the View/Edit forms could go to the link page, and from there they can re-direct to your specific edit or delete page

spencerj411 commented:
I like the sound of that.

Daniel-Jang commented:
I'd also be happy to follow Jack's suggestion. What I also thought of is that we can remove the 'remove' button from the table list view and only have the edit button. The button can then lead to an edit page, which then instead has a button to delete the course or to commit new changes.

Also, I reckon you should add text next to the edit icon ("Edit"), just so it is clear to the user.

spencerj411 commented:
That sounds good to me too. But that has brought up another point which is what about if the user wants to mass delete a set of courses?

spencerj411 commented:
What about something similar to this? This way, it will take less clicks for the user to edit/delete. Also changed actions back to text, safer to have for the mean time.
![image](https://user-images.githubusercontent.com/24206502/56276142-56e50c80-6145-11e9-9049-246dffde51ed.png)

JackSekoranja commented:
Yeah I was thinking something similar. It may even be better to remove the `edit and delete` text from the side and instead just have two buttons at the bottom saying edit selected and delete selected. I think this is a better idea since course editing and deletion is not supposed to happen very often, so it may be in our interests to make it MORE difficult to do

spencerj411 committed:
added checkboxes, changed options column to actions column, replaced imgs to text

KavsHub commented:
Hey so I was looking at this and comparing to my issue of #32. 
In the long run we're going to have heaps of courses, subplans and templates which means the table would get very long. So future thinking/early brainstorming (both you and I should look into maybe multi-pages - i.e next page etc, or a search/filter box to help the user find the relevant stuff)

spencerj411 committed:
replaced checkbox with radio buttons, mass edit/delete has been ditched for now.

spencerj411 committed:
added case where user tries to delete something when nothing is selected

Daniel-Jang committed:
Merge branches 'develop' and 'feature-course_edit_delete_base' of https://github.com/cass-degrees/CASS-Degrees-Code into feature-course_edit_delete_base

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -273,15 +273,21 @@ def manage_courses(request):
+            if rest_api is None:
                 render_properties['is_error'] = True
-                render_properties['msg'] = "Failed to delete course. An unknown error has occurred. Please try again."
+                render_properties['msg'] = 'Choose a course to delete.'
```
This is an exception, so this could be rephrased perhaps, or at least an exclamation mark?

Daniel-Jang replied:
Changed message.

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -231,6 +230,7 @@ def manage_courses(request):
         # actual_request = post_data.get('_method')
 
         if action == 'Add':
+            # courses = [{'code': course} for course in set([x['code'] for x in courses])]
```
Is this here for a reason?

Daniel-Jang replied:
No, don't think so, deleted.

j-selby commented on cassdegrees/templates/managecourses.html:
```diff
@@ -79,15 +79,52 @@
     {# Function to hide the previous tab while opening a new one #}
     <script>
+
+
```
Spacing here?

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -273,15 +273,21 @@ def manage_courses(request):
-            id_to_delete = post_data.get('id')
-
-            rest_api = requests.delete(model_api_url + id_to_delete + '/')
+            ids_to_delete= post_data.getlist('id')
```
Maybe verify these are integers? Else buggy code could potentially do unwanted stuff here (they are passed directly into the url below)

Daniel-Jang replied:
That shouldn't matter too much, the id's can't be altered by users (they are hidden input fields) and a post request with an unusual id made by something outside our Django app will be rejected since they don't have the required csrf token.

j-selby reviewed (changes requested):
No major complaints.

Daniel-Jang committed:
Completely integrated course add, edit and delete functions to list.html.
Add and Delete functions still works.
I (Daniel) also successfully implemented the "Edit courses" functionality, meaning that this user story is also complete:
https://github.com/cass-degrees/CASS-Degrees-Code/issues/31

Daniel-Jang commented:
Uhh, James, you reviewed at an unfortunate time... I have completely rewritten everything, please check the most recent commit.

With the most recent commit, I have also completed this ticket: https://github.com/cass-degrees/CASS-Degrees-Code/issues/31

j-selby commented on cassdegrees/templates/managecourses.html:
```diff
@@ -1,148 +1,76 @@
-                    <p class="text-right">
-                        <input class="btn-uni-grad btn-large" type ="submit" value="{{ action }}">
+    <div id="{{ action }}Page" {% if action == 'Delete' or render.hide_form %}hidden{% endif %}>
+        <form class="anuform" action="/manage_courses/?action={{ action }}" method="post">
```
We can't do this action stuff through POST?

Daniel-Jang replied:
Which line are you talking about? I don't understand what you are suggesting.

j-selby replied:
` action="/manage_courses/?action={{ action }}"`
Why are we mixing GET and POST?

Daniel-Jang replied:
How are we mixing GET and POST? The form most certainly states that the method is post. Adding "?action=" is not an indicator that it is a GET request.

Also, while we can create a new hidden input with the action variable from managecourses.html, there is this one problem. From list.html, there is no way to determine which button the user will press on until they do and that 'submits' the form immediately. Therefore, there is no possible way to quickly change this hypothetical hidden input value to the one we want ("Add", "Edit" or "Delete") before the form is submitted. Hence, we will have to keep it this way.

j-selby replied:
Action in a URL is a [GET parameter](https://en.wikipedia.org/wiki/Query_string). Adding action in the URL there is by definition a query string, which is what is used for GET requests. This just seems a little weird to me.

Its not too big on an issue, but something to note.

Daniel-Jang replied:
Actually, according to HTTP specifications, it is valid to have a query string for POST requests:
https://stackoverflow.com/questions/14710061/is-it-valid-to-combine-a-form-post-with-a-query-string/14710450
https://www.w3.org/Protocols/rfc2616/rfc2616.html

But I did just find out that it is generally not a good idea, so you were right to point that line out. Unfortunately, we still have to leave it this way due to the button problem :(

j-selby replied:
Na, totally fair enough - just as long as this is noted, I personally have no issue (as you said, it isn't technically incorrect).

j-selby commented on cassdegrees/templates/managecourses.html:
```diff
@@ -1,148 +1,76 @@
+    <div id="{{ action }}Page" {% if action == 'Delete' or render.hide_form %}hidden{% endif %}>
+        <form class="anuform" action="/manage_courses/?action={{ action }}" method="post">
+            {% csrf_token %}
+            {# Allow manage_courses() function to behave appropriately based on origin of request. The post request can either come from list.html or managecourses.html #}
```
IMO, this needs to not point to the origin url itself (there is a Referer header anyway, isn't there?), but as to the specific functionality that navigating from this page brings.

Daniel-Jang replied:
What are your suggested changes? I'm not sure how to proceed from your comment.

j-selby replied:
Instead of `managecourses.html`, name this option what it actually *changes*. If we are hardcoding the names of webpages here, this will get messy quick.

Daniel-Jang replied:
Yeah OK that's a good reason. Will do.

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -224,73 +223,100 @@ def manage_courses(request):
+        request_origin = post_data.get('request_origin')
+        print(request_origin)
+
+        if request_origin == 'managecourses.html':
```
I think you need to add some documentation here about what the specific differences are for this. It wasn't immediately clear to me what this suggested.

Daniel-Jang replied:
Added some comments explaining this.

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -224,73 +223,100 @@ def manage_courses(request):
-                if 'The fields code, year must make a unique set.' in rest_api.json()['non_field_errors']:
-                    render_properties['msg'] = "The course you are trying to create already exists!"
+        request_origin = post_data.get('request_origin')
+        print(request_origin)
```
Un-needed logging.

Daniel-Jang replied:
Forgot to remove them, now fixed.

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -224,73 +223,100 @@ def manage_courses(request):
+                        'name': post_data.get('name'),
+                        'units': post_data.get('units'),
+                        'offeredSem1': 'semester1' in offered_sems,
+                        'offeredSem2': 'semester2' in offered_sems
```
Might be a good idea to note if there is something in offered_sems that *isn't* handled.

Daniel-Jang replied:
Not sure what you mean by something that isn't handled?

j-selby replied:
offered_sems can potentially contain 'semester 1', 'semester 2', and (while we aren't explicitly handling it, as per our user stories), 'summer'/etc. Would just be nice to know if something unexpected is in here.

Daniel-Jang replied:
Since 'semester 1' and 'semester 2' are defined by the form (and nothing else), there is no way we will get something unexpected. Users can't interact with this either. I think your suggested checks are overkill.

j-selby replied:
Fair enough.

JackSekoranja reviewed (approved):
Looks good and works nicely. I would like to replace the radio buttons with check boxes so that the user can select and delete multiple courses at once, but I'm happy for us to create an issue to do that later so we can get this functionality out

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -224,73 +223,100 @@ def manage_courses(request):
+                    else:
+                        render_properties['is_error'] = True
+                        render_properties[
+                            'msg'] = "Failed to delete course. An unknown error has occurred. Please try again."
```
Not sure what happened here.

Daniel-Jang replied:
Whoops, me neither, fixed now.

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -224,73 +223,100 @@ def manage_courses(request):
+                else:
+                    render_properties['is_error'] = True
+                    render_properties['hide_form'] = True
+                    render_properties['msg'] = "Choose a course to edit."
```
Same feedback from before - rewording or at the very least an exclamation mark for an error.

Daniel-Jang replied:
Altered message to be "Please select a course to edit!"

j-selby commented on cassdegrees/templates/list.html:
```diff
@@ -32,35 +32,57 @@
+                        </p>
+                   </div>
+                </form>
+            {% endif %}
```
Wait, this endif ends *after* the form? That might break cases where there is no content.  (i.e `<form>` is unmatched).

Daniel-Jang replied:
Fixed, didn't see that one.

Daniel-Jang committed:
Removed residual print statements and fixed other minor issues with code as suggested by James.

Daniel-Jang committed:
Made some changes suggested by James that I missed just then.

Daniel-Jang committed:
Merge branches 'develop' and 'feature-course_edit_delete_base' of https://github.com/cass-degrees/CASS-Degrees-Code into feature-course_edit_delete_base

Daniel-Jang committed:
Refrained from using direct html file names as the value of hidden inputs.
